Source/ARTRealtime.m

```diff
@@ -771,45 +790,39 @@ - (void)onHeartbeat {
+    self.msgSerial = 0;
+    self.connection.latestMessageSerial = 0;
+    self.pendingMessageStartSerial = 0;
+}
+
 - (void)onConnected:(ARTProtocolMessage *)message {
     _renewingToken = false;
-    
-    // Resuming
-    if (_resuming) {
-        if (![message.connectionId isEqualToString:self.connection.id_nosync]) { // RTN15c3
-            [self.logger warn:@"RT:%p connection \"%@\" has reconnected, but resume failed. Reattaching any attached channels", self, message.connectionId];
-            // Reattach all channels
-            for (ARTRealtimeChannelInternal *channel in self.channels.nosyncIterable) {
-                [channel reattachWithReason:message.error callback:nil];
-            }
-            _resuming = false;
-        }
-        else if (message.error) {
-            [self.logger warn:@"RT:%p connection \"%@\" has resumed with non-fatal error \"%@\"", self, message.connectionId, message.error.message];
-            // The error will be emitted on `transition`
-        }
-        else {
-            [self.logger debug:@"RT:%p connection \"%@\" has reconnected and resumed successfully", self, message.connectionId];
-        }
-    }
-    
+
     switch (self.connection.state_nosync) {
         case ARTRealtimeConnecting: {
-            // If there's no previous connectionId, then don't reset the msgSerial
-            //as it may have been set by recover data (unless the recover failed).
-            NSString *prevConnId = self.connection.id_nosync;
-            BOOL connIdChanged = prevConnId && ![message.connectionId isEqualToString:prevConnId];
-            BOOL recoverFailure = !prevConnId && message.error;
-            if (connIdChanged || recoverFailure) {
-                [self.logger debug:@"RT:%p msgSerial of connection \"%@\" has been reset", self, self.connection.id_nosync];
-                self.msgSerial = 0;
-                self.pendingMessageStartSerial = 0;
+            if ([message.connectionId isEqualToString:self.connection.id_nosync] && !message.error) {  // RTN15c6
+                [self.logger debug:@"RT:%p connection \"%@\" has reconnected and resumed successfully", self, message.connectionId];
+                //make sure that a connection id exists for resume failure
+            } else if (self.connection.id_nosync &&
+                       ![message.connectionId isEqualToString:self.connection.id_nosync]
+                       && message.error) { // RTN15c7
+                
+                [self.logger warn:@"RT:%p connection \"%@\" has resumed with non-fatal error \"%@\"", self, message.connectionId, message.error.message];
+                [self resetSerials];
+                
+            }else if (!self.connection.id_nosync && message.error){ //recover failure
+                [self resetSerials];
+            }
+            //do not reattach on a completely new connection
+            BOOL newConnection = !self.connection.id_nosync && !message.error && !_connectionLostAt;
```

Comment:
```suggestion
            const BOOL newConnection = !self.connection.id_nosync && !message.error && !_connectionLostAt;
```

Also consider renaming to `isNewConnection`.

Reply:
5e716673a8ca4607d182bcc396ab964ba76c8895